File: gfg/2024_8_august/c++/15_add_1_to_linked_list.cpp
// TIME COMPEXITY
// O(n)

// SPACE COMPEXITY
// O(1)

// CODE

class Solution {
  public:
    Node * reverseKaryunga(Node * head){
        Node* prev = NULL;
        Node* next = NULL;
        Node* curr = head;
        
        while(curr != NULL){
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        
        return prev;
    }
    Node* addOne(Node* head) {
        head = reverseKaryunga(head);
        Node* temp = head;
        Node* tail = NULL;
        int carry = 1;
        
        while(temp && carry){
            if(temp->data == 9){
                temp->data = 0;
            }
            else{
                temp->data = temp->data + 1;
                carry = 0;
            }
            tail = temp;
            temp = temp->next;
        }
        
        // last carry ko add kro new node me. case of all 999s
        if(carry){
            tail->next = new Node(1);
        }
        head = reverseKaryunga(head);
        return head;
    }
};

File: gfg/2024_8_august/python/15_add_1_to_linked_list.py
# NOTE: In python, there is recursion limit so you can't go over 
#       that. Hence, you have to go with iterative solution only.


# CODE

class Node:
    def __init__(self, data):   # data -> value stored in node
        self.data = data
        self.next = None

class Solution:
    def reverseList(self, head):
        if not head:
            return
        curNode = head
        prevNode = head
        nextNode = head.next
        curNode.next = None
     
        while(nextNode):
            curNode = nextNode
            nextNode = nextNode.next
            curNode.next = prevNode
            prevNode = curNode
     
        return curNode
     
    def addOne(self, head):
        head = self.reverseList(head)
        k = head
        carry = 0
        prev = None
        head.data += 1
        
        while(head != None) and (head.data > 9 or carry > 0):
            prev = head
            head.data += carry
            carry = head.data // 10
            head.data = head.data % 10
            head = head.next
     
        if carry > 0:
            prev.next = Node(carry)
            
        return self.reverseList(k)